core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -785,7 +797,9 @@ private[log] class Cleaner(val id: Int,
        *   2) The message doesn't has value but it can't be deleted now.
        */
       val latestOffsetForKey = record.offset() >= foundOffset
-      val isRetainedValue = record.hasValue || retainDeletes
+      val isRetainedValue = record.hasValue ||
+                            (!batch.isDeleteHorizonSet() ||
+                             time.milliseconds() < batch.deleteHorizonMs)
```

Comment:
We probably should use the new condition only when the record batch is of the latest message format. Otherwise, we could continue to use the old condition.

Reply:
Oh, I ran tests, and spotbug main reported dodgy code. Apparently, by this point, we already know that the batch is the latest version (v_2) anyways, so we don't need to check for this.

Reply:
Hmm, how do we know at this point that batch is of V2? It seems that it could be any version.

Reply:
Alright, will push this to git then. Can't find exactly where it is checked.